### PR TITLE
Batching Feature

### LINKED ISSUE #30: Hi, any chance for adding Message Batches (beta)
It sees promising as it suits my workflow well so I would like to try.

### PR DESCRIPTION
**Pull request by WillMatthews**
Currently WIP while I implement tests, integration tests, check it actually works, refine/factor types etc.

I just thought I'd 'throw this out there' as a draft PR so that we aren't duplicating work. Happy for other input / contributions on this branch or someone else to take ownership if they're keen. I will otherwise work away on this over the next few days.

Intention to close #30

### CODE REVIEW

**Comment by liushuangls** [comment]
Thank you very much for your PR, I haven't had time to start this work recently!

**Comment by WillMatthews** [comment]
I've tested this a fair bit with the integration tests / looking at the anthropic console for batch requests. It appears to do everything vaguely sanely (I think?). I think I just want to write some tests to make codecov happy. I will do that tomorrow morning hopefully :)

File: batch.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+type ProcessingStatus string
+
+const (
+	ProcessingStatusInProgress ProcessingStatus = "in_progress"
+	ProcessingStatusCanceling  ProcessingStatus = "canceling"
+	ProcessingStatusEnded      ProcessingStatus = "ended"
+)
+
+// While in beta, batches may contain up to 10,000 requests and be up to 32 MB in total size.
```
Comment (WillMatthews) [inline batch.go:40]: Would probably be premature optimisation to verify this before sending a request.. right?

File: batch.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	// Each line in the file is a JSON object containing the result of a
+	// single request in the Message Batch. Results are not guaranteed to
+	// be in the same order as requests. Use the custom_id field to match
+	// results to requests.
+
+	Responses   []BatchResult
+	RawResponse []byte
+}
+
+// Untested - I don't know about this!
```
Comment (WillMatthews) [inline batch.go:140]: Tested, it works, but I need to refine some logic.
```suggestion
```

File: batch.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	response.RawResponse, err = io.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// for each line in the response, decode the JSON object into a MessagesResponse
+	// and append it to the Responses slice.
+	// this is tricky because the content within the response may contain newline control characters (\n)
+	// TODO: test this and make sure it works
+	for _, line := range bytes.Split(response.RawResponse, []byte("\n")) {
```
Comment (WillMatthews) [inline batch.go:179]: todo: look at this again. not overly convinced with this logic...

File: batch.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+type ListBatchRequest struct {
+	BeforeId string `json:"before_id,omitempty"`
+	AfterId  string `json:"after_id,omitempty"`
+	Limit    int    `json:"limit,omitempty"`
+}
+
+func (l ListBatchRequest) validate() error {
+	if l.Limit < 0 || l.Limit > 100 {
```
Comment (WillMatthews) [inline batch.go:212]:
```suggestion
	if l.Limit < 1 || l.Limit > 100 {
```

File: batch.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+// All times returned in RFC 3339
+type BatchResponse struct {
+	httpHeader
+
+	BatchRespCore
+}
+
+type BatchRespCore struct {
+	Id                BatchId           `json:"id"`
```
Comment (WillMatthews) [inline batch.go:58]: make id types on responses required fields!

File: batch.go (hunk 1 above)
Comment (liushuangls) [reply batch.go:40]: I think so, too

File: batch.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+	Type              BatchResponseType `json:"type"`
+	ProcessingStatus  string            `json:"processing_status"`
```
Comment (liushuangls) [inline batch.go:60]:
```suggestion
	ProcessingStatus  ProcessingStatus  `json:"processing_status"`
```

File: batch.go (hunk 3 above)
Comment (WillMatthews) [reply batch.go:179]: I'm now convinced by this logic. Evidence: https://goplay.tools/snippet/tDPm3GJVv0_s

File: batch.go (hunk 5 above)
Comment (WillMatthews) [reply batch.go:58]: not worth doing / not easily solved.

**Comment by WillMatthews** [comment]
It appears to do everything I expect it to. I now have tests for all new functions, but I would like to have some more detailed tests in the future. I think that can come with a later PR. Passes all integration tests.

Hope this is of a high enough standard!

**Review by liushuangls — APPROVED** [review]

**Comment by liushuangls** [comment]
Thank you, you have made a great contribution to the project. Would you like to be the Maintainer of the project?

**Comment by WillMatthews** [comment]
> Thank you, you have made a great contribution to the project. Would you like to be the Maintainer of the project?

Thank you very much for offering :smiley: I would be delighted to be a maintainer. I'll see if I can get #34 in too!

**Pull request merged**